Propagation soa

agarzaro opened the pull request:
+ PR to add unit test for propagation  using the pipeline flow and running in vector mode.
+ The vectorization was implemented using the VectorTypes.h library.
+ The test runs in vector mode only if the SIMD lanes are full and still have "steps" to do. Other way, like having less tracks than lanes or remaining lanes, the scalar executor is called.

amadio commented on base/inc/vectorFlow/AlignedArray.h:
```diff
@@ -0,0 +1,46 @@
+      }
+      return fArray[index];
+    }
+};
```
Maybe it would be better to write a special allocator that could be used with `std::vector<T>`, where `T` has alignment higher than 8, i.e. when `T` is a SIMD type.

agheata replied:
Indeed, we know that std::vector<SIMD_type> does not work without a special allocator, but this is why we use this new AlignedArray type, right?

agheata replied:
This is not relevant for this PR any more

amadio commented on base/inc/vectorFlow/VectorTypes.h:
```diff
@@ -3,6 +3,7 @@
 
 #include <base/Global.h>
 #include <base/Vector3D.h>
+#include <VecCore/SIMD.h>
```
Please don't include internal headers of VecCore, just `<VecCore/VecCore>`.

agheata commented on base/inc/vectorFlow/AlignedArray.h:
```diff
@@ -0,0 +1,57 @@
+    }
+
+    Data& operator[](int index) {
+      if (!(index >= 0 && index < fSize)) {
```
This is an expensive check, I would replace with an assert statement that gets compiled only in Debug mode

agarzaro commented:
Made some changes to use directly vector types Double_v instead of the AlignedArray container to avoid extra copy step. Vectorizing the operations to calculate the track step is missing, it will come in next commit.

agarzaro commented:
Added changes that also compute the step using vector types. File /test/testVectorTypesTrackPropagationPipeLine.cxx is now fully vectorized.

agarzaro commented:
Changed /test/testVectorTypesTrackPropagationPipeLine.cxx to a simple interface using vecCore masks and fixed some unnecessary extra copyings.

However, there are still some little discrepancies in the results when comparing to the scalar method results.

![image](https://user-images.githubusercontent.com/45016945/61189626-f599e000-a644-11e9-8332-e0f26f78e25f.png)

agarzaro commented:
Pushed timer for benchmarking purposes
